### PR TITLE
Add fallback if we have make a weird GC decision.

### PR DESCRIPTION
**Pull request by gbaraldi**
If something odd happens during GC (the PC goes to sleep) or a very big transient the heuristics might make a bad decision. What this PR implements is if we try to make our target more than double the one we had before we fallback to a more conservative method. This fixes the new issue @vtjnash found in https://github.com/JuliaLang/julia/issues/40644 for me.

### CODE REVIEW

**Review by vtjnash — APPROVED** [review]

**Comment by gbaraldi** [comment]
I will do a follow up PR to refactor the GC num stuff because there is currently an accounting bug somewhere and `live_bytes` gets weird results.

File: compile_commands.json
Comment (lgoettgens) [inline compile_commands.json:1]: Is this file intended to be added? It seems to contain local paths of @gbaraldi s machine.

Comment (gbaraldi) [reply compile_commands.json:1]: No, it's a mistake. I had to reinstall my os and lost my gitignore

**Review by benlorenz — COMMENTED** [review]
The currently unreliable `live_bytes` metric seems to be still in use for deciding when to do a full sweep (among other criteria):
```
    if (live_bytes > max_total_memory) {
        sweep_full = 1;
    }
```
https://github.com/JuliaLang/julia/pull/50682/files#diff-76b4d6ef0f7d9c2a6855e1ef7b0f7c94ddc68fcca2cdcd69e6d2b9007c279fdeR3273-R3275

_Edit:_ Without a heap-size-hint, cgroup restriction or the change I suggested below this would currently never trigger since the max memory is 2PB by default. But with my change this might cause the GC to always do full sweeps when when `live_bytes` is unreasonably big due to the miscount.

File: src/gc.c [hunk 1]
```diff
@@ -3612,10 +3619,10 @@ void jl_gc_init(void)
     total_mem = uv_get_total_memory();
     uint64_t constrained_mem = uv_get_constrained_memory();
     if (constrained_mem > 0 && constrained_mem < total_mem)
-        total_mem = constrained_mem;
+        jl_gc_set_max_memory(constrained_mem - 250*1024*1024); // LLVM + other libraries need some amount of memory
```
[the remark this answers is not in the document]
Comment (gbaraldi) [reply src/gc.c:3622]: Thats the idea, if you don't set an explicit hint we will just follow the heap resizing algorithm.

**Comment by oscardssmith** [comment]
is this ready to merge?

**Comment by gbaraldi** [comment]
I believe finally yes!

File: src/gc.c [hunk 2]
```diff
@@ -3612,10 +3620,10 @@ void jl_gc_init(void)
...
 #endif
     if (jl_options.heap_size_hint)
-        jl_gc_set_max_memory(jl_options.heap_size_hint);
+        jl_gc_set_max_memory(jl_options.heap_size_hint - 250*1024*1024);
```
Comment (vtjnash) [inline src/gc.c:3626]: Do we need to protect this from underflow?

**Review by vtjnash — APPROVED** [review]

File: src/gc.c (hunk 2 above)
Comment (gbaraldi) [reply src/gc.c:3626]: `jl_gc_set_max_memory` protects us fromo underflow, but it doesn't warn the user. Maybe it should?
